File: LinkedList/437-path-sum-iii/path-sum-iii.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public int pathSum(TreeNode root, int targetSum) {
       if(root == null) return 0;

       int count = countFromNode(root, targetSum);

       count+=pathSum(root.left, targetSum);
       count+=pathSum(root.right, targetSum);
       return count;

    }

    public int countFromNode(TreeNode root, long targetSum) {
        if(root == null) return 0;
        int count = 0;

        if (root.val == targetSum) count=1;
        count+= countFromNode(root.left, targetSum-root.val);
        count+= countFromNode(root.right, targetSum-root.val);
        return count;

      
    }
}

File: JavaScript/437-path-sum-iii/path-sum-iii.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def pathSum(self, root: Optional[TreeNode], targetSum: int) -> int:
        
        prefixSum = defaultdict(int)
        prefixSum[0] = 1

        def dfs(root, currSum):
            if not root:
                return 0
            
            currSum += root.val
            
            count = prefixSum[currSum-targetSum]
            prefixSum[currSum] +=1
            count+=dfs(root.left, currSum)
            count+=dfs(root.right, currSum)
            prefixSum[currSum] -=1

            return count
        if not root:
            return 0
        return  dfs(root, 0)